### PR TITLE
Disputes, Verdicts and Judgements STF Test Vectors (GP Section 10)

### PR DESCRIPTION
**Pull request by davxy**

### CODE REVIEW

**Comment by ggwpez** [comment]
Thanks for keeping the format similar to the SAFROLE tests, it helps a lot with parsing them.

- Would it be possible to also include current/previous validator keys to check that the vote signatures are valid?  
- The `age` probably indicates from which past epoch the judgement came? Currently it is only ever set to `0`.

File: disputes/disputes.asn [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+Disputes ::= SEQUENCE {
+    good SEQUENCE OF WorkReportHash,
+    bad SEQUENCE OF WorkReportHash,
+    wonky SEQUENCE OF WorkReportHash,
+    offenders SEQUENCE OF Ed25519Key
+}
+
+AvailabilityAssignment ::= SEQUENCE {
+    dummy-work-report OCTET STRING (SIZE(269)),
+    timeout U32
```
Comment (qiweiii) [inline disputes/disputes.asn:36]: is this timeslot?

**Comment by davxy** [comment]
> Thanks for keeping the format similar to the SAFROLE tests, it helps a lot with parsing them.
> 
>     * Would it be possible to also include current/previous validator keys to check that the vote signatures are valid?
> 
>     * The `age` probably indicates from which past epoch the judgement came? Currently it is only ever set to `0`. Eventually this would determine which key set to use IIUC.

@ggwpez Done

**Comment by ggwpez** [comment]
1. Would it be possible to set the `disputes` extrinsic to `null` in `progress_with_no_verdicts-1.json` to ensure that the "one or more verdicts" property of it is not violated? Like this the extrinsic is invalid, since it contains on verdicts.

Screenshot 2024-08-20 at 17 33 58

Screenshot 2024-08-20 at 17 34 22

2. Why is work report hash `0x11da6d1f761ddf9bdb4c9d6e5303ebd41f61858d0a5647a1a7bfe089bf921be9` being removed from `rho` in `progress_invalidates_avail_assignments-1.json`?

3. I think `tau` should advance by 1, otherwise `tau == tau'` in all tests.

I did not validate the markers yet, but the STF tests pass for Graymatter. Thanks!

**Comment by davxy** [comment]
> 1. Would it be possible to set the `disputes` extrinsic to `null` in `progress_with_no_verdicts-1.json` to ensure that the "one or more verdicts" property of it is not violated? Currently the extrinsic is invalid, since it contains on verdicts.


> 
>     2. Why is work report hash `0x11da6d1f761ddf9bdb4c9d6e5303ebd41f61858d0a5647a1a7bfe089bf921be9` being removed from `rho` in `progress_invalidates_avail_assignments-1.json`? I dont see any verdict or anything against it.

>     3. I think `tau` should advance by 1, otherwise `tau == tau'` in all tests.

> 
> I did not validate the markers yet, but the STF tests pass for Graymatter. Thanks! And just to clarify, the tests transition to `rho_dagger`, not to `rho_prime`, right? Since there is no availability extrinsic or anything (i appreciate this to make it easier to test).
> 
> And thanks again for your effort, the format is pretty good - hopefully other conformance tests could adapt it.

**Comment by davxy** [comment]
> 1. Would it be possible to set the `disputes` extrinsic to `null` in `progress_with_no_verdicts-1.json` to ensure that the "one or more verdicts" property of it is not violated? Currently the extrinsic is invalid, since it contains on verdicts.

Since $\textbf{v}$ is defined to belong to a set of sequences, it can include an empty sequence. If I understand correctly, this applies to all sequences in the GP (e.g., (76)). If they are empty, they are simply empty sequences.

In other words, this would be represented as `vec![]`, not `Option::None`.

Therefore, perhaps the quoted text from the GP should specify "zero or more verdicts"? @gavofyork

> 2. Why is work report hash `0x11da6d1f761ddf9bdb4c9d6e5303ebd41f61858d0a5647a1a7bfe089bf921be9` being removed from `rho` in `progress_invalidates_avail_assignments-1.json`? I dont see any verdict or anything against it.

The hash you see as the prefix of `dummy_work_report` items is actually the **work package hash**,  and not the work report hash. Specifically, it is the $h$ found in $\mathbb{S}$ by going from equation (118) to (119) to (122).

I populated that entry ($h$) with some random data to make the work report hash unique.

The report hash is then calculated by serializing the entire $w \in \mathbb{W}$ as per equation (118). If [you hash](https://emn178.github.io/online-tools/blake2b/) the `dummy_work_report` octet string from the removed entry, you'll find that its hash is `376fd230cc3710d737382e6e7cf4eaa46f3bf89589e061492e8ffdf6bb5f90d4`, which is the hash of the work report in the bad verdict.

I realize that including random data there might be misleading. I could change it to just a sequence of a single byte or something similar, or perhaps we should explicitly state what I just told in the test vector description? Or maybe both? WDYT?

> 3. I think `tau` should advance by 1, otherwise `tau == tau'` in all tests.

This is intentional because the disputes subsystem STF doesn't directly affect $\tau$. My goal is to construct the test vectors such that the post-state is strictly determined by what the GP prescribes, with timekeeping being part of section 6 (safrole, auth rotation, etc.).

> just to clarify, the tests transition to `rho_dagger`, not to `rho_prime`, right? Since there is no availability extrinsic or anything (i appreciate this to make it easier to test).

Yes, that's correct. It might be better to rename it to `rho_dagger` to make this explicit.

**Comment by ggwpez** [comment]
>The report hash is then calculated by serializing the entire w ∈ W as per equation (118). If [you hash](https://emn178.github.io/online-tools/blake2b/) the dummy_work_report octet string from the removed entry, you'll find that its hash is 376fd230cc3710d737382e6e7cf4eaa46f3bf89589e061492e8ffdf6bb5f90d4, which is the hash of the work report in the bad verdict.

Okay i see. I am just hashing the data now without trying to decode it.

>I realize that including random data there might be misleading. I could change it to just a sequence of a single byte or something similar, or perhaps we should explicitly state what I just told in the test vector description? Or maybe both? WDYT?

Yea, i was wondering since the `0x11da6d1f761ddf9bdb4c9d6e5303ebd41f61858d0a5647a1a7bfe089bf921be9` appears as work-report-hash in some other tests. So it is not really random and not unrandom either 😅 Just setting to `0...` is also fine i think.

>This is intentional because the disputes subsystem STF doesn't directly affect τ.

Yes okay. Am ignoring tau now.

**Comment by ggwpez** [comment]
>The hash you see as the prefix of dummy_work_report items is actually the work package hash, and not the work report hash. Specifically, it is the h found in S by going from equation (118) to (119) to (122).
I populated that entry (h) with some random data to make the work report hash unique.
The report hash is then calculated by serializing the entire w ∈ W as per equation (118).

Actually i went through the whole decoding logic for a Work Report now, and it seems to me that the first hash should be the `authorizer_hash`, not the work package hash from the availability spec:

Screenshot 2024-08-25 at 19 39 46

^ is the encoding equation for a Block. There is no other encoding defined for W, hence i assume this should be used?  
(either way, i can decode it and then hash it back also when using the other decoding function)

**Comment by davxy** [comment]
@ggwpez https://github.com/gavofyork/graypaper/pull/67

**Comment by vekexasia** [comment]
Hello @davxy , the https://github.com/davxy/jam-test-vectors/blob/disputes/disputes/tiny/progress_with_faults-7.json test says ( in readme) this: 

``` 
Auditor included as culprit following a verdict in his favor
```
can you clarify please? Because my test fails (correctly I guess) but the description on why that test should fail makes me question my understanding of how the verdicts/disputes system works.

In my point of view the test fails because this

https://github.com/davxy/jam-test-vectors/blob/04c2c14340abdefe6a12ab37b1470816c3147684/disputes/tiny/progress_with_faults-7.json#L38-L44

voted `true` just like the verdicts of 2/3+1 validators. so it's not really a fault. 

If this is correct there is no real formalism that prevents this in the paper besides:

1) derivative logic based on written text (not math)
2) 102 which states this:
image

Since the report hash ends up to be in `posterior psi_g` (2/3+1 validators voting yay) and since `v` is `1` then (to me) this is the reason why the test should fail. I dont see how `culprits` (which are not valued) have anything to do with the test failing?

**Comment by davxy** [comment]
@vekexasia indeed there was a typo.

The auditor, classified as an offender, had his vote aligned with the final verdict.

**Review by ggwpez — APPROVED** [review]

**Comment by xlc** [comment]
Can you explain where is this 269 coming from? https://github.com/w3f/jamtestvectors/blob/bcc43b76d5a06007022f624a786c5482b8227432/disputes/disputes.asn#L39

The minimal valid work report size that I am getting is 355 bytes (with assumption that core index is 4 bytes as I can't find the defs in GP).

I am decoding it into a WorkReport struct and if the data is invalid, well, I won't be able to decode it.

**Comment by davxy** [comment]
> Can you explain where is this 269 coming from?
> 
> https://github.com/w3f/jamtestvectors/blob/bcc43b76d5a06007022f624a786c5482b8227432/disputes/disputes.asn#L39
> 
> The minimal valid work report size that I am getting is 355 bytes (with assumption that core index is 4 bytes as I can't find the defs in GP).
> 
> I am decoding it into a WorkReport struct and if the data is invalid, well, I won't be able to decode it.

For example given this structured content:

```json
{
    "package_spec": {
        "hash": "0x30466e0ae1b05dde5249872475f6beeac368fd014b5a3413ceb32d3872143284",
        "len": 42,
        "root": "0x7af11fdaa717c398e223211842b41392f18df4bbc4ea0f4cfb972f19c7a64949",
        "segments": "0xc2fdd53ecd96faa2b4a53b5acfb113483e4b4a28d17e239824acd22fa05649b3"
    },
    "context": {
        "anchor": "0xc0564c5e0de0942589df4343ad1956da66797240e2a2f2d6f8116b5047768986",
        "state_root": "0xf6967658df626fa39cbfb6014b50196d23bc2cfbfa71a7591ca7715472dd2b48",
        "beefy_root": "0x9329de635d4bbb8c47cdccbbc1285e48bf9dbad365af44b205343e99dea298f3",
        "lookup_anchor": "0x60751ab5b251361fbfd3ad5b0e84f051ccece6b00830aed31a5354e00b20b9ed",
        "lookup_anchor_slot": 33,
        "prerequisite": null
    },
    "core_index": 3,
    "authorizer_hash": "0x022e5e165cc8bd586404257f5cd6f5a31177b5c951eb076c7c10174f90006eef",
    "auth_output": [], 
    "results": []
}
```

I get the following encoded data:

```
30466e0ae1b05dde5249872475f6beeac368fd014b5a3413ceb32d38721432842a0000007af11fdaa717c398e223211842b41392f18df4bbc4ea0f4cfb972f19c7a64949c2fdd53ecd96faa2b4a53b5acfb113483e4b4a28d17e239824acd22fa05649b3c0564c5e0de0942589df4343ad1956da66797240e2a2f2d6f8116b5047768986f6967658df626fa39cbfb6014b50196d23bc2cfbfa71a7591ca7715472dd2b489329de635d4bbb8c47cdccbbc1285e48bf9dbad365af44b205343e99dea298f360751ab5b251361fbfd3ad5b0e84f051ccece6b00830aed31a5354e00b20b9ed21000000000300022e5e165cc8bd586404257f5cd6f5a31177b5c951eb076c7c10174f90006eef0000
```

**Comment by xlc** [comment]
Screenshot 2024-09-02 at 11 50 40 AM

minimal of 1 result is required so your data is invalid

**Comment by dakk** [comment]
@davxy I don't think that the new dummy work reports are correct.

**Comment by davxy** [comment]
> @davxy I don't think that the new dummy work reports are correct.

@dakk can you elaborate? As the content is "dummy" for sure the content doesn't make sense. But maybe you're referring to the size of the raw data?

**Comment by dakk** [comment]
> > @davxy I don't think that the new dummy work reports are correct.
> 
> @dakk can you elaborate? As the content is "dummy" for sure the content doesn't make sense. But maybe you're referring to the size of the raw data?

I didn't want to send the comment without explaination (it was a mistake), I'm still debugging my decoder which works fine with CODEC tests, but fails with this: basically I'm reading 0 when it comes to read the number of workresults, but first I'm assuring that I'm following the GP correctly.

**Comment by davxy** [comment]
@dakk, as you correctly processed the vectors in the "codec" PR, I assume you're using the new JAM compact integers codec (i.e. the one in the gray paper)? These vectors are still utilizing the old SCALE compact integer encoding. I will regenerate them today I promise 😅 , and then we can retry.

**Comment by dakk** [comment]
> @dakk, as you correctly processed the vectors in the "codec" PR, I assume you're using the new JAM compact integers codec (i.e. the one in the gray paper)? These vectors are still utilizing the old SCALE compact integer encoding. I will regenerate them today I promise 😅 , and then we can retry.

Oh gosh, I think that is the problem  😅 I'll wait for it, thank you

**Comment by dakk** [comment]
@davxy now my tests are passing, but you should also update the json files since dummy_work_report is encoded differently between the binary and the json.

**Comment by dakk** [comment]
Hi @davxy , do you had a chance to check discrepancies between the binary and the json format regarding the dummy_work_report? This is an example of difference between the twos:

progress_invalidates_avail_assignments-1.json

rho - prestate - first dummy work report 
```0x11da6d1f761ddf9bdb4c9d6e5303ebd41f61858d0a5647a1a7bfe089bf921be9000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c0302010400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002a00000000000000000c010203```

progress_invalidates_avail_assignments-1.bin
rho - prestate - first dummy work report
```0x11da6d1f761ddf9bdb4c9d6e5303ebd41f61858d0a5647a1a7bfe089bf921be9000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c0302010400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002a000000000000000003010203```

**Comment by davxy** [comment]
@dakk currenty the JSON is constructed using the old codec (which is why you're seeing the dummy work report blob encoded in the older format). This happens because we first encode using the "classic" scale codec and then convert it to JSON using the `scalecodec` Python library: https://github.com/polkascan/py-scale-codec.

To produce a blob compatible with the new codec, we would need to fork the `py-scale-codec` and introduce the new compact encoding (more or less here: https://github.com/polkascan/py-scale-codec/blob/f676638998ce612d9f0b38b135f77a95eacd51ce/scalecodec/types.py#L31).

It shouldn't be too difficult, and since we're using that library for test vectors, I may implement this at some point (unless someone has already done it :-) ).

**Comment by dakk** [comment]
> 't be too difficult, and since we're using that library for test vectors, I may implement this a

Got it (y) 
No, it is not a blocker; I disabled those test on jampy for the moment.

**Comment by jaymansfield** [comment]
Just wanted to mention "progress_with_bad_signatures-2" seems to match two different errors I'm not sure if that was intended.

Depending on which order the implementers code their validation rules it could result in "bad signature", or "already judged" since the target exists in psi_b pre-state already.

**Comment by bloppan** [comment]
> Just wanted to mention "progress_with_bad_signatures-2" seems to match two different errors I'm not sure if that was intended.
> 
> Depending on which order the implementers code their validation rules it could result in "bad signature", or "already judged" since the target exists in psi_b pre-state already.

Yes, and also can match with culprits not sorted unique error.

**Comment by davxy** [comment]
This is another reason why it’s better not to be picky about the error codes returned by the STF. Invalid inputs may trigger different errors depending on the order of checks, which is unspecified in the GP.

That said, I agree that it’s **very helpful to have test vectors that don’t rely on a specific order of checks**. In future revisions, I’ll fix these signaled vectors to make the returned errors not ambiguous. The feedback gathered here is not being overlooked

**Comment by davxy** [comment]
Temporarily closed in favor of https://github.com/w3f/jamtestvectors/pull/28

**Pull request closed without merging**